### PR TITLE
Generative improv

### PR DESCRIPTION
**Pull request by Giuseppe5**

### CODE REVIEW

File: src/brevitas_examples/stable_diffusion/main.py [hunk 1]
```diff
@@ -5,79 +5,152 @@
...
-    if args.export_target:
-        assert args.weight_quant_format == 'int', "Currently only integer quantization supported for export."
+    dtype = getattr(torch, args.dtype)
 
-    # Select dtype
-    if args.float16:
-        dtype = torch.float16
-    else:
-        dtype = torch.float32
+    calibration_prompts = CALIBRATION_PROMPTS
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/main.py:141]: Consider restructuring this.

File: src/brevitas_examples/stable_diffusion/main.py [hunk 2]
```diff
@@ -107,33 +203,47 @@ def main(args):
     # Extract list of layers to avoid
     blacklist = []
     for name, _ in pipe.unet.named_modules():
-        if 'time_emb' in name or 'conv_in' in name:
-            blacklist.append(name)
+        if 'time_emb' in name and not args.quantize_time_emb:
+            blacklist.append(name.split('.')[-1])
+        if 'conv_in' in name and not args.quantize_conv_in:
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/main.py:208]: Remove this if not needed.

File: src/brevitas_examples/stable_diffusion/main.py [hunk 3]
```diff
@@ -147,22 +257,50 @@ def weight_bit_width(module):
...
+        def input_bit_width(module):
+            if isinstance(module, nn.Linear):
+                return args.linear_input_bit_width
+            elif isinstance(module, nn.Conv2d):
+                return args.conv_input_bit_width
+            else:
+                raise RuntimeError(f"Module {module} not supported.")
+
+        input_kwargs = dict()
+        if args.linear_input_bit_width is None or args.conv_input_bit_width is None:
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/main.py:270]: Apply your fix and we re-review.

File: src/brevitas_examples/stable_diffusion/main.py [hunk 4]
```diff
@@ -178,52 +316,153 @@ def input_bit_width(module):
...
+                    test_latents=latents,
+                    guidance_scale=args.guidance_scale)
+
+        if args.load_checkpoint is not None:
+            with load_quant_model_mode(pipe.unet):
+                pipe = pipe.to('cpu')
+                pipe.unet.load_state_dict(torch.load(args.load_checkpoint, map_location='cpu'))
+                pipe = pipe.to(args.device)
+        elif not args.dry_run:
+            if (args.linear_input_bit_width is not None or
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/main.py:367]: Fix this.

File: src/brevitas_examples/stable_diffusion/main.py [hunk 5]
```diff
@@ -178,52 +316,153 @@ def input_bit_width(module):
...
+            fid.update(float_images_values, real=True)
+            fid.update(quant_images_values, real=False)
+            print(f"FID: {float(fid.compute())}")
 
     if args.export_target:
         # Move to cpu and to float32 to enable CPU export
-        if not (args.float16 and args.export_cuda_float16):
-            pipe.unet.to('cpu').to(torch.float32)
+        if not (dtype == torch.float16 and args.export_cuda_float16):
+            pipe.unet.to('cpu').to(dtype)
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/main.py:465]: Check if this breaks for ONNX export.

File: src/brevitas_examples/stable_diffusion/main.py [hunk 6]
```diff
@@ -248,6 +487,13 @@ def input_bit_width(module):
                 export_manager = StdQCDQONNXManager
                 export_manager.change_weight_export(export_weight_q_node=args.export_weight_q_node)
             export_onnx(pipe, trace_inputs, output_dir, export_manager)
+        if args.export_target == 'torch':
+            if args.weight_quant_granularity == 'per_group':
+                export_manager = BlockQuantProxyLevelManager
+            else:
+                export_manager = TorchQCDQManager
+                export_manager.change_weight_export(export_weight_q_node=args.export_weight_q_node)
+            export_torch_export(pipe, trace_inputs, output_dir, export_manager)
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/main.py:496]: Remove this.

File: src/brevitas_examples/stable_diffusion/main.py [hunk 7]
```diff
@@ -389,8 +702,50 @@ def input_bit_width(module):
...
     add_bool_arg(
         parser, 'export-cuda-float16', default=False, help='Export FP16 on CUDA. Default: Disabled')
+    add_bool_arg(
+        parser,
+        'use-mlperf-inference',
+        default=False,
+        help='Evaluate FID score with MLPerf pipeline. Default: False')
+    add_bool_arg(
+        parser,
+        'use-ocp',
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/main.py:719]: Decide if you want to keep this argument or not.

File: src/brevitas_examples/stable_diffusion/main.py [hunk 8]
```diff
@@ -389,8 +702,50 @@ def input_bit_width(module):
...
+        parser,
+        'dry-run',
+        default=False,
+        help='Generate a quantized model without any calibration. Default: Disabled')
+    add_bool_arg(
+        parser,
+        'quantize-time-emb',
+        default=True,
+        help='Quantize time embedding layers. Default: True')
+    add_bool_arg(
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/main.py:737]: Remove these 3 options.

File: src/brevitas_examples/stable_diffusion/sd_quant/utils.py [hunk 9]
```diff
@@ -3,8 +3,106 @@
 SPDX-License-Identifier: MIT
 """
 
+from contextlib import contextmanager
+
 import torch
 
+from brevitas.export.common.handler.base import BaseHandler
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/sd_quant/utils.py:10]: Remove all of this.

**Review by nickfraser — CHANGES_REQUESTED** [review]
We just need a re-review of the bit related to setting the input quantization.

File: src/brevitas_examples/stable_diffusion/sd_quant/export.py [hunk 10]
```diff
@@ -3,31 +3,113 @@
...
+    layer_dict['input_zp'] = input_zp.numpy().tolist()
+    layer_dict['input_zp_shape'] = input_zp.shape
+    layer_dict['input_zp_dtype'] = str(torch.int8)
+    layer_dict['weight_scale'] = weight_scale.numpy().tolist()
+    nelems = layer.weight.shape[0]
+    weight_scale_shape = [nelems] + [1] * (layer.weight.data.ndim - 1)
+    layer_dict['weight_scale_shape'] = weight_scale_shape
+    layer_dict['weight_zp'] = weight_zp.numpy().tolist()
+    layer_dict['weight_zp_shape'] = weight_scale_shape
+    layer_dict['weight_zp_dtype'] = str(torch.uint8)
```
Comment (nickfraser) [inline src/brevitas_examples/stable_diffusion/sd_quant/export.py:52]: Should be `int8`?

**Review by nickfraser — CHANGES_REQUESTED** [review]

File: src/brevitas_examples/stable_diffusion/sd_quant/export.py [hunk 11]
```diff
@@ -3,31 +3,113 @@
...
+        output_scale = layer.output_quant.export_handler.symbolic_kwargs[
+            'dequantize_symbolic_kwargs']['scale'].data
+
+        layer_dict['output_scale'] = output_scale.numpy().tolist()
+        layer_dict['output_scale_shape'] = output_scale.shape
+    layer_dict['input_scale'] = input_scale.numpy().tolist()
+    layer_dict['input_scale_shape'] = input_scale.shape
+    layer_dict['input_zp'] = input_zp.numpy().tolist()
+    layer_dict['input_zp_shape'] = input_zp.shape
+    layer_dict['input_zp_dtype'] = str(torch.int8)
```
Comment (Giuseppe5) [inline src/brevitas_examples/stable_diffusion/sd_quant/export.py:45]: Restore to uint8

**Pull request merged**